MREUnityRuntime/MREUnityRuntimeLib/Core/Components/AnimationComponent.cs

```diff
@@ -122,128 +196,267 @@ void AddKeyframe(MWAnimationKeyframe keyframe)
+        internal void Interpolate(
+            ActorPatch finalFrame,
+            string animationName,
+            float duration,
+            float[] curve,
+            bool enabled,
+            Action onCompleteCallback)
         {
-            var animation = GetOrLookUpUnityAnimationComponent();
-            if (animation != null)
+            // Ensure duration is in range [0...n].
+            duration = Math.Max(0, duration);
+
+            // If the curve is malformed, fallback to linear.
+            if (curve.Length != 4)
             {
-                if (animation[animationName] != null)
+                curve = new float[] { 0, 0, 1, 1 };
+            }
+
+            // Are we patching the transform?
+            bool animateTransform = finalFrame.Transform != null && finalFrame.Transform.IsPatched();
+            var finalTransform = finalFrame.Transform;
+
+            // What parts of the transform are we animating?
+            bool animatePosition = animateTransform && finalTransform.Position != null && finalTransform.Position.IsPatched();
+            bool animateRotation = animateTransform && finalTransform.Rotation != null && finalTransform.Rotation.IsPatched();
+            bool animateScale = animateTransform && finalTransform.Scale != null && finalTransform.Scale.IsPatched();
+
+            // Ensure we have a well-formed rotation quaternion.
+            if (animateRotation)
+            {
+                bool wellFormed =
+                    finalTransform.Rotation.X.HasValue &&
+                    finalTransform.Rotation.Y.HasValue &&
+                    finalTransform.Rotation.Z.HasValue &&
+                    finalTransform.Rotation.W.HasValue;
+
+                // If quaternion is malformed, fallback to the identity.
+                if (!wellFormed)
                 {
-                    _hasRootMotion[animationName] = hasRootMotion.HasValue && hasRootMotion.Value;
-                    int scalar = paused.HasValue ? paused.Value ? 0 : 1 : 1;
-                    float time = animationTime ?? 0;
-                    animation[animationName].enabled = scalar > 0 ? true : false;
-                    animation[animationName].speed = 1f * scalar;
-                    animation[animationName].weight = 1f * scalar;
-                    animation[animationName].time = time;
+                    finalTransform.Rotation = new QuaternionPatch(Quaternion.identity);
                 }
             }
-        }
 
-        internal void StopAnimation(string animationName, float? animationTime)
-        {
-            var animation = GetUnityAnimationComponent();
-            if (animation)
+            // Create the sampler to calculate ease curve values.
+            var sampler = new CubicBezier(curve[0], curve[1], curve[2], curve[3]);
+
+            bool LerpFloat(out float dest, float start, float? end, float t)
```

Comment:
Move all these local named functions to the end of Interpolate, so all the immediate code is together.

Reply:
Done.